docs/hooks.md

```diff
@@ -0,0 +1,141 @@
+  [x]) ;; x = 1
+
+;; 2nd update
+(uix/use-effect
+  (fn [] (prn x))
+  [x]) ;; x = 1, didn't change, do nothing
+
+;; 3rd update
+(uix/use-effect
+  (fn [] (prn x)) ;; prn 2
+  [x]) ;; x = 2, `x` has changed, rerun the hook
+```
+
+That works as expected for primitive values like numbers and strings, but what about Clojure's maps and vectors that can be compared by value?
+
+```clojure
+(= {:x 1} {:x 1}) ;; true
+```
+
+Two maps equal by value but with different identity are considered to be not equal when React compares dependencies. This is happening because React is a JS library that is not aware of Clojure's equality algorithm, instead React is comparing values by identity `===` or `Object.is`. Thus it's important to think about what type of values you are passing as dependencies of a hook. It applies to JS as well, objects and arrays are still compared by reference.
+
+```clojure
+;; 1st update
+(let [a {:x 1}]
+  (uix/use-effect
+    (fn [] (prn a)) ;; executed
+    [a])))
+
+;; 2nd update
+(let [a {:x 1}]
+  (uix/use-effect
+    (fn [] (prn a)) ;; executed as well, since `a` is a new map that was created during render
+    [a])))
+
+;; 3rd update
+(let [a {:x 1}]
+  (uix/use-effect
+    (fn [] (prn a)) ;; same, still executed for the same reason
+    [a])))
+```
+
+### What about other Clojure's primitives?
+
+In addition to JavaScript's primitive types like `Number` or `String`, ClojureScript has keywords, symbols and UUIDs that are represented as JS objects when compiled into JS and thus fall into the same trap with equality check. To make things simpler UIx automatically stringifies those three types when passed in the dependency vector.
+
+## Return value in effect hooks
+
+The _setup_ function, that is passed into one of the effect hooks, requires the return value to be either a function (that will be called on _cleanup_) or `js/undefined`. Otherwise React will throw an error saying that it got something else.
+
+```clojure
+(react/useEffect
+  (fn []
+    :keyword) ;; returning `:keyword` here will throw
+  #js [])
+```
+
+Now in ClojureScript, when an expression returns _nothing_, it actually returns `nil`, which is compiled into `null` in JS. Since `null` is neither a function, nor `undefined`, React will throw as well.
+
+```clojure
+(react/useEffect
+  (fn []
+    (when false (prn :x))) ;; returns `nil` and thus React throws
+  #js [])
+```
+
+Thus when using React hooks directly you'd have to explicitly return `js/undefined` in such cases.
+
+```clojure
+(react/useEffect
+  (fn []
+    (when false (prn :x))
+    js/undefined)
+  #js [])
+```
+
+This complication is also handled by UIx and if the return value is not a function it will automatically return `js/undefined`. However keep in mind that since in Clojure the last expression is always returned implicitly, you still have to make sure that an expression doesn't return a function accidentally, because it's going to be executed in _cleanup_ phase of an effect hook.
```

Comment:
```suggestion
This complication is also handled by UIx and if the return value is not a function it will automatically return `js/undefined`. However keep in mind that, since in Clojure the last expression is always returned implicitly, you still have to make sure the hook doesn't return a function accidentally, because it's going to be executed in its _cleanup_ phase.
```